pkg/kubernetes/kubernetes.js

```diff
@@ -34,97 +34,239 @@ define([
+                })
                 .done(function(data) {
-                    var resp = { };
-                    $(self).triggerHandler("changed");
-                    if (data) {
-                        try {
-                            resp = JSON.parse(data);
-                        } catch(ex) {
-                            console.warn("etcd parse exception", ex, data);
-                        }
+                    var cancelled = req && req.cancelled;
+                    req = null;
+                    if (!stopping && !cancelled) {
+                        if (!waited)
+                            console.warn("watching kubernetes " + type + " didn't block");
+                        else
+                            start_watch();
                     }
-                    var nindex;
-                    if (resp.node && resp.node.modifiedIndex)
-                        nindex = resp.node.modifiedIndex;
-                    else if (resp.prevNode && resp.prevNode.modifiedIndex)
-                        nindex = resp.prevNode.modifiedIndex;
-                    if (nindex !== undefined) {
-                        nindex++;
-                        notify(nindex);
-                    } else {
-                        timeout = window.setTimeout(function() { notify(); }, 2000);
-                    }
-                })
-                .fail(function(ex) {
-                    request = null;
-                    console.warn("etcd: " + ex.message);
                 });
+            requested = true;
         }
 
-        self.poke = function poke() {
-            if (!request)
-                notify();
-        };
+        start_watch();
 
-        self.close = function close() {
-            if (request)
-                request.close();
+        self.stop = function stop() {
+            stopping = true;
+            if (req)
+                req.close();
         };
-
-        notify();
     }
 
     function KubernetesClient() {
         var self = this;
 
         var api = cockpit.http(8080);
-        var first = true;
+        var watches = [ ];
+
+        /*
+         * Here we create KubernetesWatch object, and a property
+         * definition for each type of object and hook them together.
+         */
+        function bind(type, items) {
+            var flat = null;
+            var timeout;
+
+            /* Always delay the update event a bit */
+            function trigger() {
+                flat = null;
+                if (!timeout) {
+                    timeout = window.setTimeout(function() {
+                        timeout = null;
+                        $(self).triggerHandler(type, self[type]);
+                    }, 100);
+                }
+            }
 
-        self.nodes = [ ];
-        self.pods = [ ];
-        self.services = [ ];
-        self.replicationcontrollers = [];
+            function update(items, item) {
```

Comment:
See line 253 below. But I'll rename this function, it's confusing to have both a callback named ```update``` and a function named ```update``` when they are not the same thing.